add v3 header

jaimenorman opened the pull request:
add vertex information to parse_ascii, to allow for the v3 header to be converted/skimmed

raymondEhlers commented on jetscape_analysis/analysis/reader/parse_ascii.py:
```diff
@@ -165,7 +168,29 @@ def _parse_header_line_format_unspecified(line: str) -> HeaderInfo:
     # save some string comparisons.
     info: Union[HeaderInfo, CrossSection]
-    if (len(values) == 19 and values[1] == "Event") or (len(values) == 17 and values[1] == "Event"):
+    if (len(values) == 17 and values[1] == "Event"):
```
I don't think this will maintain comparability with the v2 header. The reason that the original if statement has 17 or 19 is because some value is optional (I don't recall which one). I'm also confused how you could have added values, but the expected length is now shorter?

raymondEhlers replied:
Actually, looking at this some more, I guess this isn't precisely compatible w/ the `_parse_header_line_format_v2` function, which isn't ideal. I know this isn't precisely your problem, but this additional makes the issue clear, so maybe you can help me figure this out? My questions:

1. Do we actually need to try to parse the unrecognized header for all cases? I guess there may be some v2 headers w/out the header version, so we need the v2 there, but that shouldn't be the case for v3. So may it doesn't need to be here at all?
2. Can we just call the decided parsers for the various headers here if we pass the if statement? I guess we run the check twice, but maybe that's a reasonable cost in the case of the version not being defined?
3. Why is the length that we pick out for the v2 parser significantly different in the two versions? Is is just a typo? Or does the v2 parser not work? (There are tests, so I guess it should work)

jaimenorman replied:
Hi Raymond:
1: perhaps not - my tests indeed ran the '_parse_header_line_format_v3' part. In fact, when is the unrecognised header code ran? If it is not needed, then I could remove
2: I guess that this could be done - I could make the changes if you agree?
3. I think this is the addition of pt hat, which adds 2 characters, is this what you meant? I'll take a look at the tests when I get a chance

raymondEhlers replied:
You can run the existing tests with `pytest -l tests`, which **should** be enough to determine if any of the unspecified format is broken. I think the easiest path forward would be for you to undo the changes you made to the unspecified parser, we don't mess with \#2, and then we can just rely on the v3 parser for your changes

raymondEhlers commented on jetscape_analysis/analysis/reader/parse_ascii.py:
```diff
@@ -270,6 +347,7 @@ def _parse_header_line_format_v2(line: str) -> HeaderInfo:
 # Register header parsing functions
 _file_format_version_to_header_parser = {
     2: _parse_header_line_format_v2,
+    3: _parse_header_line_format_v3,
```
Can you please try to add some tests? If I recall correctly, they're rudimentary, but they're a bit useful. See them here: `JETSCAPE-analysis/tree/master/tests)/test_parsing.py` (there are some instructions on how they're generated in the comments)

jaimenorman replied:
yes, I can take a look at adding the tests

raymondEhlers reviewed (commented):
I think the base functionality looks okay. I don't want to hold you up, but it would be good if we could clarify a few things here before merging. The details are slowly coming back to me, so I'm hoping that it should be pretty straightforward

jaimenorman commented:
Hi Raymond,
I did what you suggested - i.e. removed the unspecified header parsing and just added that for v3 (also updating so that it can run with and without pt_hat). I tested and all looks OK, could you merge?
I understood from your comment that the tests wouldn't need changing as well, but let me know if I should do anything here.
Thanks!

raymondEhlers commented:
Hi Jaime, you could enable v3 for the tests, but good enough for now. Merging

jaimenorman commented:
Thanks Raymond, could you also recreate the container when you get a chance?
On the tests - I thought that the tests were only for the unspecified setup, but if not I can add one for v3 as well.

On 29 Aug 2022, at 23:48, Raymond Ehlers ***@***.******@***.***>> wrote:



Merged #21<https://github.com/jdmulligan/JETSCAPE-analysis/pull/21> into master.

—
Reply to this email directly, view it on GitHub<https://github.com/jdmulligan/JETSCAPE-analysis/pull/21#event-7282643099>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AGKGX4QV6RGRX5LZNU6RVATV3U42HANCNFSM56EVW25Q>.
You are receiving this because you authored the thread.Message ID: ***@***.***>

raymondEhlers commented:
New container is ready. It's on OSN as usual, local v3.8 .

For the testing, I guess it's a bit ambiguous. I suppose if [here](https://github.com/jdmulligan/JETSCAPE-analysis/blob/8177b07e5f3bcd1e13410680051efa34f8238e8f/tests/test_parsing.py#L70) is generalized to not use parser_v1_header_v1 as the reference array, it would probably be a more meaningful test. I don't really have the bandwidth to think through this further right now, so if you see the opportunity to add a meaningful test, go for it. Otherwise, just skip it

jaimenorman commented:
Thanks! It is working well with the new changes (and less failing jobs which I think you mentioned would be the case).


On 31 Aug 2022, at 13:05, Raymond Ehlers ***@***.******@***.***>> wrote:



New container is ready. It's on OSN as usual, local v3.8 .

For the testing, I guess it's a bit ambiguous. I suppose if here<https://github.com/jdmulligan/JETSCAPE-analysis/blob/8177b07e5f3bcd1e13410680051efa34f8238e8f/tests/test_parsing.py#L70> is generalized to not use parser_v1_header_v1 as the reference array, it would probably be a more meaningful test. I don't really have the bandwidth to think through this further right now, so if you see the opportunity to add a meaningful test, go for it. Otherwise, just skip it

—
Reply to this email directly, view it on GitHub<https://github.com/jdmulligan/JETSCAPE-analysis/pull/21#issuecomment-1232848062>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AGKGX4WI6YI4TFPT5RSLYK3V35DBZANCNFSM56EVW25Q>.
You are receiving this because you authored the thread.Message ID: ***@***.***>
